sdk/src/metrics/export/periodic_exporting_metric_reader.cc

```diff
@@ -96,12 +102,77 @@ bool PeriodicExportingMetricReader::CollectAndExportOnce()
       break;
     }
   } while (status != std::future_status::ready);
+  bool notify_force_flush = is_force_flush_pending_.exchange(false, std::memory_order_acq_rel);
+  if (notify_force_flush)
+  {
+    is_force_flush_notified_.store(true, std::memory_order_release);
+    force_flush_cv_.notify_one();
+  }
+
   return true;
 }
 
 bool PeriodicExportingMetricReader::OnForceFlush(std::chrono::microseconds timeout) noexcept
 {
-  return exporter_->ForceFlush(timeout);
+  std::unique_lock<std::mutex> lk_cv(force_flush_m_);
+  is_force_flush_pending_.store(true, std::memory_order_release);
+  auto break_condition = [this]() {
+    if (IsShutdown())
+    {
+      return true;
+    }
+
+    // Wake up the worker thread once.
+    if (is_force_flush_pending_.load(std::memory_order_acquire))
+    {
+      is_force_wakeup_background_worker_.store(true, std::memory_order_release);
+      cv_.notify_one();
+    }
+    return is_force_flush_notified_.load(std::memory_order_acquire);
+  };
+
+  timeout = opentelemetry::common::DurationUtil::AdjustWaitForTimeout(
+      timeout, std::chrono::microseconds::zero());
+  std::chrono::steady_clock::duration timeout_steady =
+      std::chrono::duration_cast<std::chrono::steady_clock::duration>(timeout);
+  if (timeout_steady <= std::chrono::steady_clock::duration::zero())
+  {
+    timeout_steady = std::chrono::steady_clock::duration::max();
+  }
+
+  bool result = false;
+  while (!result && timeout_steady > std::chrono::steady_clock::duration::zero())
+  {
+    // When is_force_flush_notified_.store(true) and force_flush_cv_.notify_all() is called
+    // between is_force_flush_pending_.load() and force_flush_cv_.wait(). We must not wait
+    // for ever
+    std::chrono::steady_clock::time_point start_timepoint = std::chrono::steady_clock::now();
+    result = force_flush_cv_.wait_for(lk_cv, export_interval_millis_, break_condition);
+    timeout_steady -= std::chrono::steady_clock::now() - start_timepoint;
+  }
+
+  // If it will be already signaled, we must wait util notified.
+  // We use a spin lock here
+  if (false == is_force_flush_pending_.exchange(false, std::memory_order_acq_rel))
+  {
+    for (int retry_waiting_times = 0;
+         false == is_force_flush_notified_.load(std::memory_order_acquire); ++retry_waiting_times)
+    {
+      opentelemetry::common::SpinLockMutex::fast_yield();
+      if ((retry_waiting_times & 127) == 127)
+      {
+        std::this_thread::yield();
+      }
+    }
+  }
+
+  is_force_flush_notified_.store(false, std::memory_order_release);
+  if (result)
+  {
+
+    result = exporter_->ForceFlush(timeout);
```

Comment:
Is it better to use `timeout_steady` if it's greater than zero here? We have already wait some time before and the left timeout should be less.

Reply:
Good point. Have changed it to use `timeout_steady` now.